syntactic/src/main/scala/syntactic_checker/CheckResult.scala

```diff
@@ -0,0 +1,29 @@
+package syntactic_checker
+
+/**
+ * Result of checking a program
+ */
+sealed trait CheckResult[+V <: Violation]
+
+object CheckResult {
+
+  /**
+   * syntactic_checker.CheckResult meaning that the program complies to the specified list of features
+   */
+  case object Valid extends CheckResult[Nothing]
+
+  /**
+   * syntactic_checker.CheckResult meaning that the program contains forbidden trees
+   *
+   * @param violations the forbidden trees
+   */
+  case class Invalid[+V <: Violation](violations: List[V]) extends CheckResult[V]
```

Comment:
Yes exactly. I find it better to use distinct implementations for both types of violations (since in `BlacklistViolation` we also have to report the violated rule but not in `WhitelistViolation`), and I don't see any other way but parametrizing `CheckResult` to do it.